cli/up/build.go

```diff
@@ -1,436 +1,103 @@
-				break loop
-			}
-
-			fmt.Printf("Pushing image for %s:\n", service)
-			img := images[service]
-			err := cmd.pushServiceImage(img.id, img.name)
-			if err != nil {
-				return nil, errors.WithContext(fmt.Sprintf("push %s", img.name), err)
-			}
-
-		case <-timer.C:
-			if !progressPrinting {
-				pp = util.NewProgressPrinter(os.Stdout, "Waiting for base images to be uploaded")
-				go pp.Run()
-				progressPrinting = true
-			}
-		}
-	}
-
-	return svcToImageName, nil
-}
-
-func (cmd *up) getServicesToPush(services composeTypes.Services) (toPush composeTypes.Services) {
-	// We only need to check to see if cached images are pushed. If an image
-	// isn't cached, it isn't built yet, so we're going to push it after
-	// building.
-	cachedImageNames := []string{}
-	svcToCachedImage := map[string]string{}
-	for _, svc := range services {
-		id, ok := cmd.getCachedImage(svc.Name)
-		if ok {
-			imageName := getImageName(cmd.imageNamespace, svc.Name, id)
-			cachedImageNames = append(cachedImageNames, imageName)
-			svcToCachedImage[svc.Name] = imageName
-		}
-	}
-
-	// Before pushing an image, we first check to see if the remote manifest
-	// already exists. This is more efficient than doing a full image push
-	// because we don't compare each individual layer.
-	pushedImages := getPushedImages(cachedImageNames, cmd.auth.AuthToken)
-	for _, svc := range services {
-		if _, ok := pushedImages[svcToCachedImage[svc.Name]]; !ok {
-			toPush = append(toPush, svc)
-		}
-	}
-
-	return toPush
-}
-
-// prePushBaseImages initiates the pre-push requests for the given services. It
-// returns a channel which sends the names of services as their prepush finishes
-// or fails.
-func (cmd *up) prePushBaseImages(services composeTypes.Services) (<-chan string, error) {
-	// We should send each service on this channel exactly once, then exit.
-	readyToPush := make(chan string, len(services))
-
-	// Keep track of which services still need to be sent over the channel, to
-	// make sure they are all eventually sent. We will delete from this map as
-	// we send.
-	// This is not thread-safe, so only one thread should use it at a time.
-	needToSend := map[string]struct{}{}
-	for _, svc := range services {
-		needToSend[svc.Name] = struct{}{}
-	}
-	ready := func(service string) {
-		if _, ok := needToSend[service]; !ok {
-			panic("tried to send same service over the channel twice")
-		}
-		delete(needToSend, service)
-		readyToPush <- service
-	}
-
-	tagRequests := []*cluster.TagImageRequest{}
-
-	for _, svc := range services {
-		dockerfile := svc.Build.Dockerfile
-		if dockerfile == "" {
-			dockerfile = "Dockerfile"
-		}
-
-		dockerfilePath := filepath.Join(filepath.Dir(cmd.composePath),
-			svc.Build.Context, dockerfile)
-
-		baseImageName, err := cmd.getBaseImage(dockerfilePath)
-		if err != nil {
-			log.WithError(err).WithField("service", svc.Name).Debug("Failed to get base image from Dockerfile")
-			ready(svc.Name)
+		if _, ok := builtImages[svc.Name]; ok {
```

Comment:
I think it's more obviously correct to check `cmd.alwaysBuild` here as well. I know returning the empty map has the  same effect, but we can consider that an _optimization_, while checking here is the _implementation_